src/codegen/llvm.zig

```diff
@@ -7501,6 +7503,56 @@ pub const FuncGen = struct {
         return self.wip.bin(.udiv, lhs, rhs, "");
     }
 
+    fn airDivCeil(self: *FuncGen, inst: Air.Inst.Index, fast: Builder.FastMathKind) !Builder.Value {
+        const o = self.dg.object;
+        const mod = o.module;
+        const bin_op = self.air.instructions.items(.data)[inst].bin_op;
+        const lhs = try self.resolveInst(bin_op.lhs);
+        const rhs = try self.resolveInst(bin_op.rhs);
+        const inst_ty = self.typeOfIndex(inst);
+        const scalar_ty = inst_ty.scalarType(mod);
+
+        if (scalar_ty.isRuntimeFloat()) {
+            const result = try self.buildFloatOp(.div, fast, inst_ty, 2, .{ lhs, rhs });
+            return self.buildFloatOp(.ceil, fast, inst_ty, 1, .{result});
+        } else if (scalar_ty.isSignedInt(mod)) {
+            const inst_llvm_ty = try o.lowerType(inst_ty);
+            const bit_size_minus_one = try o.builder.splatValue(inst_llvm_ty, try o.builder.intConst(
+                inst_llvm_ty.scalarType(&o.builder),
+                inst_llvm_ty.scalarBits(&o.builder) - 1,
+            ));
+            const zero = try o.builder.zeroInitValue(inst_llvm_ty);
+            const one = try o.builder.splatValue(inst_llvm_ty, try o.builder.intConst(inst_llvm_ty.scalarType(&o.builder), 1));
+
+            const quotient = try self.wip.bin(.sdiv, lhs, rhs, "");
+            const rem = try self.wip.bin(.srem, lhs, rhs, "");
+
+            const quotient_sign = try self.wip.bin(.xor, lhs, rhs, "");
+            // quotient_sign_mask has all bits set to 1 if the result is negative, or 0
+            // otherwise.
+            const quotient_sign_mask = try self.wip.bin(.ashr, quotient_sign, bit_size_minus_one, "");
+            // correction_if_inexact is 0 if the result is negative, 1 otherwise.
+            const correction_if_inexact = try self.wip.bin(.add, quotient_sign_mask, one, "");
+            const is_rem_nonzero = try self.wip.icmp(.ne, rem, zero, "");
+            const correction = try self.wip.select(fast, is_rem_nonzero, correction_if_inexact, zero, "");
+
+            return self.wip.bin(.@"add nsw", quotient, correction, "");
+        } else if (scalar_ty.isUnsignedInt(mod)) {
+            const inst_llvm_ty = try o.lowerType(inst_ty);
+            const zero = try o.builder.zeroInitValue(inst_llvm_ty);
+            const one = try o.builder.splatValue(inst_llvm_ty, try o.builder.intConst(inst_llvm_ty.scalarType(&o.builder), 1));
+
+            const quotient = try self.wip.bin(.udiv, lhs, rhs, "");
+            const rem = try self.wip.bin(.urem, lhs, rhs, "");
+            const is_non_zero = try self.wip.icmp(.ne, rem, zero, "");
+            const correction = try self.wip.select(fast, is_non_zero, one, zero, "");
+
+            return try self.wip.bin(.@"add nuw", quotient, correction, "");
+        } else {
+            @panic("Unexpected scalar type");
```

Comment:
This `@panic` will not be optimized out in release builds. If it's never supposed to happen it's fine to just use `unreachable`.
```suggestion
            unreachable;
```

Reply:
I don't think it's a good idea to use `unreachable` in such cases. If the assumption that the code is unreachable is incorrect, or becomes incorrect due to a future change, then instead of an easy to diagnose `@panic` you get potentially subtle undefined behaviour.

In return, you get a performance improvement that is going to be insignificant a lot of the time.

The compiler does already use `unreachable` a lot in cases like this, but I suspect it contains a lot of undefined behaviour as a result (for example: I'm pretty sure an `unreachable` in `AirDivFloor` in `wasm/CodeGen.zig` would be reachable by calling `@divFloor` with 128 bit integers if the compiler didn't hit an assertion failure - which itself triggers undefined behaviour in release mode - before getting there).

Reply:
Since writing this, I've realised / remembered that Zig with `-ReleaseFast` treats integer overflow as undefined behaviour, so avoiding potential undefined behaviour in `-ReleaseFast` is a bit of a lost cause. In any case, Zig has an established convention of using `unreachable` rather than `@panic` in these cases.

I've changed all the `@panic`s I added to `unreachable`s.